### PR TITLE
✨ Mailbox

### PR DESCRIPTION
**Pull request by EvolveArt**
Completes 1st milestone of the Starknet implementation.

### CODE REVIEW

File: src/contracts/client/mailboxclient.cairo [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    impl Upgradeable of IUpgradeable<ContractState> {
+        fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
+            self.ownable.assert_only_owner();
+            self.upgradeable._upgrade(new_class_hash);
+        }
+    }
+
+
+    #[abi(embed_v0)]
+    impl IMailboxClientImpl of IMailboxClient<ContractState> {
```
Comment (EvolveArt) [inline src/contracts/client/mailboxclient.cairo:60]: I can't see the `assert_only_mailbox` modifier ?

File: src/contracts/client/mailboxclient.cairo [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        fn _is_delivered(self: @ContractState, _id: u256) -> bool {
+            let mailbox_address = self.mailbox.read();
+            let mailbox = IMailboxDispatcher { contract_address: mailbox_address };
+            mailbox.delivered(_id)
+        }
+
+        fn _dispatch(
+            self: @ContractState,
```
Comment (EvolveArt) [inline src/contracts/client/mailboxclient.cairo:94]: I think that's a mutable function so it should be `ref self: ContractState`

File: src/contracts/client/mailboxclient.cairo [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        fn quote_dispatch(
+            self: @ContractState,
+            _destination_domain: u32,
+            _recipient: ContractAddress,
+            _message_body: Bytes,
+            _hook_metadata: Option<Bytes>,
+            _hook: Option<ContractAddress>
+        ) {
```
Comment (EvolveArt) [inline src/contracts/client/mailboxclient.cairo:113]:
```suggestion
        ) -> u256 {
```

File: src/contracts/client/mailboxclient.cairo (hunk 2 above)
Comment (EvolveArt) [reply src/contracts/client/mailboxclient.cairo:94]: Also the solidity impl returns `bytes32` please explain the reasoning behind using `u256` in comments

File: src/contracts/libs/message.cairo [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            nonce: 0_u32,
+            origin: 0_u32,
+            sender: contract_address_const::<0>(),
+            destination: 0_u32,
+            recipient: contract_address_const::<0>(),
+            body: BytesTrait::new_empty(),
+        }
+    }
+
+    fn format_message(message: Message) -> u256 {
```
Comment (EvolveArt) [inline src/contracts/libs/message.cairo:33]: explain the formatting method in comment

File: src/contracts/mailbox.cairo [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+        upgradeable: UpgradeableComponent::Storage,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        DefaultIsmSet: DefaultIsmSet,
+        DefaultHookSet: DefaultHookSet,
+        RequiredHookSet: RequiredHookSet,
+        Process: Process,
```
Comment (EvolveArt) [inline src/contracts/mailbox.cairo:63]: this is not possible yet ?
```suggestion
        Process,
```

File: src/contracts/mailbox.cairo [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+            let message = build_message(
+                self, _destination_domain, _recipient_address, _message_body.clone()
+            );
+            let required_hook_address = self.required_hook.read();
+            let required_hook = IPostDispatchHookDispatcher {
+                contract_address: required_hook_address
+            };
+            let hook = IPostDispatchHookDispatcher { contract_address: hook_address };
+            required_hook.quote_dispatch(hook_metadata.clone(), message.clone())
+                - hook.quote_dispatch(hook_metadata, message)
```
Comment (EvolveArt) [inline src/contracts/mailbox.cairo:394]:
```suggestion
                + hook.quote_dispatch(hook_metadata, message)
```

File: src/interfaces.cairo [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    ROUTING,
+    AGGREGATION,
+    LEGACY_MULTISIG,
+    MERKLE_ROOT_MULTISIG,
+    MESSAGE_ID_MULTISIG,
+    NULL, // used with relayer carrying no metadata
+    CCIP_READ,
+}
+
+pub const HYPERLANE_VERSION: u8 = 3;
```
Comment (EvolveArt) [inline src/interfaces.cairo:33]: let's use this constant in `message.cairo` then

File: src/contracts/mailbox.cairo (hunk 5 above)
Comment (JordyRo1) [reply src/contracts/mailbox.cairo:63]: It is not possible for now.

File: src/contracts/client/mailboxclient.cairo (hunk 2 above)
Comment (JordyRo1) [reply src/contracts/client/mailboxclient.cairo:94]: We are modifying the storage of the underlying mailbox contract, not the mailboxclient one, that's why. In solidity, the message formatting returns a bytes32, as a result of the keccak hashing process. In our case, we are working with the same keccak hashing returns a u256, so we had to make some changes. I will write a small article with all the significant changes like this one and detail why I considered these changes.

File: src/contracts/client/mailboxclient.cairo (hunk 1 above)
Comment (JordyRo1) [reply src/contracts/client/mailboxclient.cairo:60]: The assertion assert_only_mailbox is not used in the MailboxClient implementation, so no real need to add it for this contract.

File: src/utils/keccak256.cairo [hunk 8]
```diff
@@ -0,0 +1,8 @@
+use core::integer::u128_byte_reverse;
+
+/// Reverse the endianness of an u256
+pub fn reverse_endianness(value: u256) -> u256 {
```
Comment (EvolveArt) [inline src/utils/keccak256.cairo:4]: add unit test

**Pull request merged**